### PR TITLE
Restart DB process test

### PR DESCRIPTION
**Pull request by marceloneppel**
# Issue
* Jira issues: [DPE-543](https://warthogs.atlassian.net/browse/DPE-543)
* The PostgreSQL charm should self heal the workload when restarting the DB process.

# Solution
* Add a test that restarts the DB processes (from Patroni and from PostgreSQL) and later check that the workload recovers itself from that situation.

# Context
* This test can also be run on an existing cluster.

* Code from `tests/integration/ha_tests/helpers.py` was copied from the VM charm, so you can focus only on the `send_signal_to_process ` helper function, which has some differences (it uses the same approach that MySQL K8S operatoe is using due to some errors and/or timeouts when trying to use `juju ssh`).

* The test was implemented on `tests/integration/ha_tests/test_self_healing.py`.

* `.github/workflows/release.yaml` was updated to reuse the tests from `.github/workflows/ci.yaml`.

# Testing
* The test was added on `tests/integration/ha_tests/test_self_healing.py`. The test restarts both Patroni and PostgreSQL OS process, one at a time.

# Release Notes
* Add restart db process test.

### CODE REVIEW

File: tests/integration/ha_tests/test_self_healing.py [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+) -> None:
+    # Locate primary unit.
+    app = await app_name(ops_test)
+    primary_name = await get_primary(ops_test, app)
+
+    # Start an application that continuously writes data to the database.
+    await start_continuous_writes(ops_test, app)
+
+    # Freeze the database process.
+    await send_signal_to_process(ops_test, primary_name, process, "SIGTERM")
```
Comment (WRFitch) [inline tests/integration/ha_tests/test_self_healing.py:64]: Is it worth trying this with SIGKILL as well?

File: tests/integration/ha_tests/helpers.py [hunk 2]
```diff
@@ -0,0 +1,337 @@
...
+    Args:
+        ops_test: OpsTest instance.
+    """
+
+    def get_host_ip(host: str) -> str:
+        # Translate the pod hostname to an IP address.
+        model = ops_test.model.info
+        client = Client(namespace=model.name)
+        pod = client.get(Pod, name=host.split(".")[0])
+        return pod.status.podIP
```
Comment (MiaAltieri) [inline tests/integration/ha_tests/helpers.py:116]: neat

File: tests/integration/ha_tests/helpers.py [hunk 3]
```diff
@@ -0,0 +1,337 @@
...
+            # list provided by other members.
+            assert member_ips == {
+                get_host_ip(member["host"]) for member in cluster_info.json()["members"]
+            }, "members report different lists of cluster members."
+        else:
+            member_ips = {get_host_ip(member["host"]) for member in cluster_info.json()["members"]}
+    return member_ips
+
+
+async def get_master_start_timeout(ops_test: OpsTest) -> Optional[int]:
```
Comment (MiaAltieri) [inline tests/integration/ha_tests/helpers.py:134]: change master naming

File: tests/integration/ha_tests/helpers.py [hunk 4]
```diff
@@ -0,0 +1,337 @@
...
+
+    Returns:
+        string with the password stored on the peer relation databag.
+    """
+    # Can retrieve from any unit running unit, so we pick the first.
+    for attempt in Retrying(stop=stop_after_attempt(len(ops_test.model.applications[app].units))):
+        with attempt:
+            unit_name = (
+                ops_test.model.applications[app].units[attempt.retry_state.attempt_number - 1].name
+            )
```
Comment (MiaAltieri) [inline tests/integration/ha_tests/helpers.py:181]: If the PostGres service has been paused on a unit and then you ask that unit for the password will this still work? 

If not might want to change which unit you're grabbing to be any unit except the one that was paused

File: tests/integration/ha_tests/helpers.py [hunk 5]
```diff
@@ -0,0 +1,337 @@
...
+
+    try:
+        for attempt in Retrying(stop=stop_after_delay(60), wait=wait_fixed(3)):
+            with attempt:
+                with psycopg2.connect(
+                    connection_string
+                ) as connection, connection.cursor() as cursor:
+                    cursor.execute("SELECT COUNT(number) FROM continuous_writes;")
+                    secondary_writes = cursor.fetchone()[0]
+                    assert secondary_writes == expected_writes
```
Comment (MiaAltieri) [inline tests/integration/ha_tests/helpers.py:253]: is this just checking the last write that was written to the database or all writes? I don't know SQL well enough 😅

File: tests/integration/ha_tests/conftest.py [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+    # Deploy the continuous writes application charm if it wasn't already deployed.
+    async with ops_test.fast_forward():
+        if await app_name(ops_test, APPLICATION_NAME) is None:
+            charm = await ops_test.build_charm("tests/integration/ha_tests/application-charm")
+            await ops_test.model.deploy(charm, application_name=APPLICATION_NAME)
+            await ops_test.model.wait_for_idle(status="active", timeout=1000)
+    yield
+    # Clear the written data at the end.
+    action = await ops_test.model.units.get(f"{APPLICATION_NAME}/0").run_action(
+        "clear-continuous-writes"
```
Comment (MiaAltieri) [inline tests/integration/ha_tests/conftest.py:28]: does this also stop the writes?

File: tests/integration/ha_tests/test_self_healing.py [hunk 7]
```diff
@@ -0,0 +1,105 @@
...
+@pytest.mark.parametrize("process", DB_PROCESSES)
+async def test_restart_db_process(
+    ops_test: OpsTest, process: str, continuous_writes, master_start_timeout
+) -> None:
+    # Locate primary unit.
+    app = await app_name(ops_test)
+    primary_name = await get_primary(ops_test, app)
+
+    # Start an application that continuously writes data to the database.
+    await start_continuous_writes(ops_test, app)
```
Comment (MiaAltieri) [inline tests/integration/ha_tests/test_self_healing.py:61]: if you wanted you could just put this in the fixture but up to you

**Review by MiaAltieri — APPROVED** [review]
Approved, minor comments and questions only

File: .github/workflows/ci.yaml [hunk 8]
```diff
@@ -97,6 +98,24 @@ jobs:
...
 
+  integration-test-microk8s-ha-self-healing:
+    name: Integration tests for high availability self healing (microk8s)
+    needs:
+      - lint
+      - unit-test
+    runs-on: ubuntu-latest
+    steps:
+      - name: Checkout
+        uses: actions/checkout@v2
```
[the remark this answers is not in the document]
Comment (marceloneppel) [reply .github/workflows/ci.yaml]: Thanks for this info. Updated on https://github.com/canonical/postgresql-k8s-operator/pull/59/commits/5b1322e4dd05a5a8a9640012538065622d3d3dc3.

File: tests/integration/ha_tests/helpers.py [hunk 9]
```diff
@@ -0,0 +1,337 @@
...
+        await ops_test.model.applications[app].add_unit(count=1)
+        await ops_test.model.wait_for_idle(apps=[app], status="active", timeout=1000)
+
+    # Load Kubernetes configuration to connect to the cluster.
+    config.load_kube_config()
+
+    # Send the signal.
+    pod_name = unit_name.replace("/", "-")
+    command = f"pkill --signal {signal} -f {process}"
+    response = stream(
```
[the remark this answers is not in the document]
Comment (marceloneppel) [reply tests/integration/ha_tests/helpers.py:247]: In some tests, like the freeze db process test (that will be the next to be implemented), some errors are returned (or the call hangs) when using `ops_test` available options.

File: tests/integration/ha_tests/test_self_healing.py (hunk 1 above)
Comment (marceloneppel) [reply tests/integration/ha_tests/test_self_healing.py:60]: `SIGKILL` will be used in the kill db process test. That test wasn't implemented yet due to some issues on pebble.

File: tests/integration/ha_tests/conftest.py [hunk 10]
```diff
@@ -0,0 +1,41 @@
...
+    )
+    await action.wait()
+
+
+@pytest.fixture()
+async def master_start_timeout(ops_test: OpsTest) -> None:
+    """Temporary change the master start timeout configuration."""
```
[the remark this answers is not in the document]
Comment (marceloneppel) [reply tests/integration/ha_tests/conftest.py:56]: In this specific case `master` is part of the parameter name, so I think it's better to keep it, only in this case, to not cause any confusion about which parameter is being changed.

File: tests/integration/ha_tests/helpers.py [hunk 11]
```diff
@@ -0,0 +1,337 @@
...
+class MemberNotListedOnClusterError(Exception):
+    """Raised when a member is not listed in the cluster."""
+
+
+class MemberNotUpdatedOnClusterError(Exception):
+    """Raised when a member is not yet updated in the cluster."""
+
+
+class ProcessError(Exception):
+    pass
```
[the remark this answers is not in the document]
Comment (marceloneppel) [reply tests/integration/ha_tests/helpers.py]: Good catch. Added on https://github.com/canonical/postgresql-k8s-operator/pull/59/commits/9f03ffa27e9c4261e60d24777f770a5f5b06c2a7.

File: tests/integration/ha_tests/helpers.py [hunk 12]
```diff
@@ -0,0 +1,337 @@
...
+    """
+    status = await ops_test.model.get_status()
+    for app in ops_test.model.applications:
+        if application_name in status["applications"][app]["charm"]:
+            return app
+
+    return None
+
+
+async def change_master_start_timeout(ops_test: OpsTest, seconds: Optional[int]) -> None:
```
[the remark this answers is not in the document]
Comment (marceloneppel) [reply tests/integration/ha_tests/helpers.py:51]: Same as above.

File: tests/integration/ha_tests/helpers.py (hunk 3 above)
Comment (marceloneppel) [reply tests/integration/ha_tests/helpers.py:126]: Same as above.

File: tests/integration/ha_tests/conftest.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (marceloneppel) [reply tests/integration/ha_tests/conftest.py:56]: This is a Patroni configuration.

File: tests/integration/ha_tests/test_self_healing.py (hunk 7 above)
Comment (marceloneppel) [reply tests/integration/ha_tests/test_self_healing.py]: Great suggestion! Moved this code to a fixture on https://github.com/canonical/postgresql-k8s-operator/pull/59/commits/9b9576c1cbe81736868b524682707532ea030fe4.

File: tests/integration/ha_tests/conftest.py (hunk 6 above)
Comment (marceloneppel) [reply tests/integration/ha_tests/conftest.py:49]: No. I updated the code to stop it in the fixture on https://github.com/canonical/postgresql-k8s-operator/pull/59/commits/9b9576c1cbe81736868b524682707532ea030fe4.

File: tests/integration/ha_tests/helpers.py (hunk 4 above)
Comment (marceloneppel) [reply tests/integration/ha_tests/helpers.py]: This mechanism was already handling that situation (getting the info from a different unit when a unit is paused).

However, I updated the code on https://github.com/canonical/postgresql-k8s-operator/pull/59/commits/ac65418440fda0771a07d2563a5b6bc6fc329f87 (along with some other changes) to have a better implementation like we have in the VM charm and other charms.

File: tests/integration/ha_tests/helpers.py (hunk 5 above)
Comment (marceloneppel) [reply tests/integration/ha_tests/helpers.py:211]: This gets the total amount of numbers written (to compare with the last write that was returned after stopping the writes from the continuous writes charm).

**Review by MiaAltieri — APPROVED** [review]
awesome work Marcelo :)

**Review by taurus-forever — APPROVED** [review]
Superb!

File: tests/integration/test_tls.py [hunk 13]
```diff
@@ -50,7 +50,7 @@ async def test_mattermost_db(ops_test: OpsTest) -> None:
         )
         # Deploy TLS Certificates operator.
         config = {"generate-self-signed-certificates": "true", "ca-common-name": "Test CA"}
-        await ops_test.model.deploy(TLS_CERTIFICATES_APP_NAME, channel="edge", config=config)
+        await ops_test.model.deploy(TLS_CERTIFICATES_APP_NAME, channel="beta", config=config)
```
Comment (taurus-forever) [inline tests/integration/test_tls.py:53]: BTW, https://charmhub.io/tls-certificates-operator channel beta has version 13, while edge is on 18.
Is Beta version good enough for us (just a question)?

**Review by WRFitch — APPROVED** [review]

File: tests/integration/test_tls.py (hunk 13 above)
Comment (taurus-forever) [reply tests/integration/test_tls.py:53]: JFYI, reported the fresh issue: https://github.com/canonical/tls-certificates-operator/issues/23

**Pull request merged**